### PR TITLE
Minor front-end refactoring

### PR DESCRIPTION
**Pull request by absentabyss**
Front-end now processes requests and updates the content instead of redirecting the user to raw JSON pages.

### CODE REVIEW

File: cgi-bin/sendConsoleCommand.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 def core(response):
     if utils.server_is_alive():
-        command = base64.b64decode(commandEncoded).decode('utf-8')
+        command = base64.b64decode(command_encoded).decode('utf-8')
```
Comment (pidement) [inline cgi-bin/sendConsoleCommand.py:11]: why the change of this variable only here?

File: index.html [hunk 2]
```diff
@@ -1,400 +1,287 @@
...
+                />
+            </div>
+            <div>
+                Resource pack:
+                <input type="text" name="resource-pack" id="resource-pack" />
+            </div>
+            <div>
+                <input type="submit" value="Save" /><label
+                    id="server-properties-success"
+                ></label>
```
Comment (pidement) [inline index.html]: This indentation it's a little confusing.

File: index.html [hunk 3]
```diff
@@ -1,400 +1,287 @@
...
+                <input type="submit" name="submit-level-name" value="Set" />
+            </form>
+        </div>
+        Upload level folder in ZIP format:
+        <form id="upload-level" method="POST" enctype="multipart/form-data">
+            File: <input name="file" type="file" />
+            <br />
+            <input name="submit" type="submit" value="Upload Level" /><label
+                id="upload-level-success"
+            ></label>
```
Comment (pidement) [inline index.html]: same. I guess that it's the default vim indentation. Can we change it?

File: index.html [hunk 4]
```diff
@@ -1,400 +1,287 @@
...
+        <script type="text/javascript" src="./scripts/versionManagement.js"></script>
+        <script
+            type="text/javascript"
+            src="./scripts/startStopServer.js"
+        ></script>
+        <script
+            type="text/javascript"
+            src="./scripts/serverProperties.js"
+        ></script>
+        <script type="text/javascript" src="./scripts/publicIp.js"></script>
```
Comment (pidement) [inline index.html]: The same here.
```></script>```

File: scripts/utils.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+function newButton(buttonId, text) {
+    var button = `<button id="${buttonId}" type="button">${text}</button>`;
+    return button;
+}
+
+function refreshElement(element) {
+    var new_element = element.cloneNode(true);
+    element.parentNode.replaceChild(new_element, element);
+}
```
Comment (pidement) [inline scripts/utils.js]: why this function it's needed?

File: scripts/levelManagement.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
+function configureDownloadLevelButton() {
+    var downloadLevelButton = document.getElementById(
+        "selected-level-download"
+    );
+    refreshElement(downloadLevelButton);
+    var downloadLevelButton = document.getElementById(
+        "selected-level-download"
+    );
```
Comment (pidement) [inline scripts/levelManagement.js]: this element it's defined twice.

File: scripts/levelManagement.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        var request = new XMLHttpRequest();
+        request.open("GET", url, true);
+        request.onload = function () {
+            var res = JSON.parse(request.responseText);
+            if (res["success"] === "true") {
+                window.open(res["url"]);
+            }
+        };
+        request.onerror = function () {};
+        request.send(); // create FormData from form that triggered event
```
Comment (pidement) [inline scripts/levelManagement.js]: why not use ```httpGetAsync``` here ?

File: scripts/levelManagement.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        var request = new XMLHttpRequest();
+        request.open("GET", url, true);
+        request.onload = function () {
+            var res = JSON.parse(request.responseText);
+            if (res["success"] === "true") {
+                updateLevelModule();
+            }
+        };
+        request.onerror = function () {};
+        request.send(); // create FormData from form that triggered event
```
Comment (pidement) [inline scripts/levelManagement.js]: Why do not use ```httpGetAsync``` here ?

File: scripts/levelManagement.js [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        event.preventDefault();
+    });
+}
+
+function getSelectedLevel(responseJSON) {
```
Comment (pidement) [inline scripts/levelManagement.js]: I think that we can rename this function to something like ```prepareSelectedLevel``` or ```setSelectedLever```

File: scripts/levelManagement.js [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    var level = JSON.parse(responseJSON)["level-name"];
+    document.getElementById("selected-level").innerHTML = level;
+    configureDownloadLevelButton();
+    configureDeleteLevelButton();
+}
+
+function getAvailableLevels(responseJSON) {
```
Comment (pidement) [inline scripts/levelManagement.js]: The same here. This function modifies some elements and don't return a value.

File: scripts/levelManagement.js [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+        // request failed
+    };
+
+    request.send(new FormData(event.target)); // create FormData from form that triggered event
+    event.preventDefault();
+}
+
+function updateLevelModule() {
+    httpGetAsync("/cgi-bin/getCurrentLevel.py", getSelectedLevel);
+    httpGetAsync("/cgi-bin/getLevelsList.py", getAvailableLevels);
```
Comment (pidement) [inline scripts/levelManagement.js]: Another possible name for the functions mentioned above are ```updateSelectedLevel``` and ```updateAvailableLevels```

File: scripts/utils.js [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+ *     request.onerror = function () {
+ *         // request failed
+ *     };
+ *
+ *     request.send(new FormData(event.target)); // create FormData from form that triggered event
+ *     event.preventDefault();
+ * }
+ * // It can be attached to a form with the following command.
+ * document.getElementById(formId).addEventListener("submit", formSubmit);
+ */
```
Comment (pidement) [inline scripts/utils.js]: remove commented code

File: scripts/versionManagement.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+            var request = new XMLHttpRequest();
+            request.open("GET", url, true);
+            request.onload = function () {
+                var res = JSON.parse(request.responseText);
+                if (res["success"] === "true") {
+                    updateCurrentVersion();
+                }
+            };
+            request.onerror = function () {};
+            request.send(); // create FormData from form that triggered event
```
Comment (pidement) [inline scripts/versionManagement.js]: Why not use ```httpGetAsync``` ?

**Review by pidement — CHANGES_REQUESTED** [review]

File: cgi-bin/sendConsoleCommand.py (hunk 1 above)
Comment (absentabyss) [reply cgi-bin/sendConsoleCommand.py:11]: Snake case is consistently used within Python scripts.

File: index.html (hunk 2 above)
Comment (absentabyss) [reply index.html]: Indentation is handled by Prettier. We can override it if you want.

File: scripts/utils.js (hunk 5 above)
Comment (absentabyss) [reply scripts/utils.js]: I think it's explained in the Commit. Elements can have listeners, this is a way to clear listeners that have been previously added to the element (and more). There's another method that removes specific listeners, but some listeners have a Lambda callback so I couldn't use that method.

File: scripts/levelManagement.js (hunk 9 above)
Comment (absentabyss) [reply scripts/levelManagement.js]: Maybe `refreshSelectedLevelModule`?

File: scripts/levelManagement.js (hunk 10 above)
Comment (absentabyss) [reply scripts/levelManagement.js]: Could you specify the issue? I don't see a problem with not returning anything under the current implementation.

File: scripts/levelManagement.js (hunk 6 above)
Comment (absentabyss) [reply scripts/levelManagement.js]: That's because I wasn't sure if refreshElement destroyed the old element or not. I've had problems with similar functions in the past, so I did that just to be safe.

File: scripts/utils.js (hunk 12 above)
Comment (absentabyss) [reply scripts/utils.js]: It's a preset for sending forms through JavaScript. The URL is different for each form, and I don't know how to pass URL as an argument on event listeners.

File: index.html (hunk 3 above)
Comment (absentabyss) [reply index.html]: Refer to my first reply on the issue.

File: index.html (hunk 4 above)
Comment (absentabyss) [reply index.html]: Refer to my first reply on the issue.

File: scripts/versionManagement.js (hunk 13 above)
Comment (absentabyss) [reply scripts/versionManagement.js]: Whoops!

File: scripts/levelManagement.js (hunk 7 above)
Comment (absentabyss) [reply scripts/levelManagement.js]: Whoops!

File: scripts/levelManagement.js (hunk 8 above)
Comment (absentabyss) [reply scripts/levelManagement.js]: Whoops!

File: scripts/levelManagement.js (hunk 11 above)
Comment (absentabyss) [reply scripts/levelManagement.js]: I agree.

File: cgi-bin/sendConsoleCommand.py (hunk 1 above)
Comment (pidement) [reply cgi-bin/sendConsoleCommand.py:11]: Yes I know, but why the rename It's only on this variable? before this PR the variable it's not used?

File: index.html (hunk 2 above)
Comment (pidement) [reply index.html]: I think that
```html
<input type="submit" value="Save" />
<label id="server-properties-sucess"></label>
```
it's better than:
```html
<input type="submit" value="Save" /><label
id="server-properties-sucess"
></label>
```
Can we make this change?

File: index.html (hunk 3 above)
Comment (pidement) [reply index.html]: I'll put this as resolved.

File: cgi-bin/sendConsoleCommand.py (hunk 1 above)
Comment (absentabyss) [reply cgi-bin/sendConsoleCommand.py:11]: It's probably a fix from a previous refactoring. When I moved a lot of functions into utils.py I think I introduced a Bug by creating a command_encoded variable and forgot to rename this one instance.

File: scripts/levelManagement.js (hunk 10 above)
Comment (pidement) [reply scripts/levelManagement.js]: This functions edit some elements of the UI. So I think that it is better a name like ```updateAvailableLevels``` or something like that.

File: index.html (hunk 2 above)
Comment (absentabyss) [reply index.html]: Yes. What should we do with:
```html
></script>
```
? Personally, I don't mind it.

Comment (absentabyss) [reply index.html]: Script opening and closing tags are now in the same line.

**Review by pidement — APPROVED** [review]
👏 Nice Work!
